man/lxsession.1
'\" t
.\"     Title: lxsession
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: March  2, 2008
.\"    Manual: General Commands Manual
.\"    Source: https://lxde.org
.\"  Language: English
.\"
.TH "LXSESSION" "1" "March 2, 2008" "https://lxde\&.org" "General Commands Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
lxsession \- lightweight X session manager
.SH "SYNOPSIS"
.HP \w'\fBlxsession\fR\ 'u
\fBlxsession\fR [\fB\fIOPTIONS\fR\fR...]
.SH "DESCRIPTION"
.PP
\fBlxsession\fR
is a program that provides an X session manager\&. It is the default X session manager for LXDE\&.
.PP
A session manager automatically starts a set of applications and sets up a working desktop environment\&. Additionally, when a user logs out, the session manager remembers which applications were in use and restarts them the next time the user logs in\&.
.SH "OPTIONS"
.PP
\fB\-a\fR, \fB\-\-noautostart\fR
.RS 4
disable autostart of applications (window\-manager mode only)
.RE
.PP
\fB\-c\fR, \fB\-\-compatibility \fR\fB\fIname\fR\fR
.RS 4
specify a compatibility mode for settings (only
\fIrazor\-qt\fR
is supported)
.RE
.PP
\fB\-e\fR, \fB\-\-de \fR\fB\fIname\fR\fR
.RS 4
name of the desktop environnement (\fILXDE\fR
is the default)
.RE
.PP
\fB\-n\fR, \fB\-\-noxsettings\fR
.RS 4
disable Xsettings daemon support
.RE
.PP
\fB\-r\fR, \fB\-\-reload\fR
.RS 4
reload configurations (for Xsettings daemon)
.RE
.PP
\fB\-s\fR, \fB\-\-session \fR\fB\fIname\fR\fR
.RS 4
name of the desktop session profile
.RE
.SH "AUTHORS"
.PP
Ying\-Chun Liu
.PP
This manual page was written by paulliu
<grandpaul@gmail\&.com>\&.
.SH "COPYRIGHT"
.PP
\(co 2008 paulliu
.PP
Permission is granted to copy, distribute and/or modify this document under the terms of the
GNU
General Public License, Version 2 any later version published by the Free Software Foundation\&.
.SH "SEE ALSO"
.PP
\fBlxpanel\fR(1),
\fBlxsession\-logout\fR(1)


debian/NEWS
lxsession (0.4.3-1) unstable; urgency=low

  * New option introduced since version 0.4:
    -e NAME	specify name of DE, such as LXDE, GNOME, or XFCE.

  If you gets crash during startup after upgrade from 0.3.8 to 0.4,
  you may hitting the same problem mentioned at #575961. You may need
  to specific the new option to avoid this problem.

 -- Andrew Lee (李健秋) <ajqlee@debian.org>  Sun, 11 Apr 2010 22:33:23 +0800


src/lxsession.c
/*
 *      lxsession.c
 *
 *      Copyright 2008 PCMan <pcman.tw@gmail.com>
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; either version 2 of the License, or
 *      (at your option) any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License
 *      along with this program; if not, write to the Free Software
 *      Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *      MA 02110-1301, USA.
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <glib.h>
#include <gtk/gtk.h>

#include <unistd.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <string.h>
#include <stdlib.h>

#include "lxsession.h"
#include "xevent.h"
#include "settings-daemon.h"
#include "xdg-autostart.h"
#include "default-apps.h"
#include "clipboard.h"


static gboolean no_settings = FALSE; /* disable settings daemon */
static gboolean reload_settings = FALSE; /* reload settings daemon */
static gboolean no_autostart = FALSE; /* no autostart */

gboolean config_changed = FALSE; /* Set to TRUE if need to save conf file */

static GMainLoop* main_loop = NULL;
static const char *display_name = NULL;
char* window_manager = NULL; /* will be accessed by settings-daemon.c */

/* name of environment variables */
/* Disable not used
static char sm_env[] = "SESSION_MANAGER";
*/
static char display_env[] = "DISPLAY";
static char pid_env[] = "_LXSESSION_PID";

static char prog_name[]="lxsession";
static char autostart_filename[]="autostart";

const char *session_name = NULL;
const char* de_name = NULL;

static void start_session(GKeyFile* kf);

static void sig_term_handler ( int sig )
{
    g_main_loop_quit(main_loop);
}

void lxsession_quit()
{
    g_main_loop_quit(main_loop);
}

static void register_signals()
{
    /* Ignore SIGPIPE */
    signal( SIGPIPE, SIG_IGN );

#if 0
    action.sa_handler = g_child_watch_signal_handler;
    sigemptyset (&action.sa_mask);
    action.sa_flags = SA_RESTART | SA_NOCLDSTOP;
    sigaction (SIGCHLD, &action, NULL);
#endif

    /* If we get a SIGTERM, do logout */
    signal( SIGTERM, sig_term_handler );
}

GKeyFile* load_session_config( const char* config_filename )
{
    const gchar* const *dirs = g_get_system_config_dirs();
    const gchar* const *dir;
    GKeyFile* kf = g_key_file_new();
    char* filename;
    gboolean ret;

    /* load user-specific session config */
    filename = g_build_filename( g_get_user_config_dir(), prog_name, session_name, config_filename, NULL );
    ret = g_key_file_load_from_file(kf, filename, 0, NULL);
    g_free( filename );

	if( ! ret ) /* user specific file is not found */
	{
		/* load system-wide session config files */
		for( dir = dirs; *dir; ++dir )
		{
			filename = g_build_filename( *dir, prog_name, session_name, config_filename, NULL );
			ret = g_key_file_load_from_file(kf, filename, 0, NULL);
			g_free( filename );
			if(ret)
				break;
		}
	}

	if( G_UNLIKELY(!ret) )
	{
		g_key_file_free(kf);
		return NULL;
	}
	return kf;
}

/*
 * system wide default config is /etc/xdg/lxsession/SESSION_NAME/desktop.conf
 * system wide default apps are listed in /etc/xdg/lxsession/SESSION_NAME/autostart
 */
void start_session(GKeyFile* config)
{
    const gchar* const *dirs = g_get_system_config_dirs();
    const gchar* const *dir;
    char* filename;

    /* run window manager first */
    g_message("Will launch window manager \n");
    config_changed = app_command_window_manager(config);
    g_message("Window manager launched \n");

    /* run the panel if it's set */
    if (g_key_file_get_string( config, "Session", "panel/program", NULL))
    {
        app_command_panel(config);
    }

    /* run the screensaver if it's set */
    if (g_key_file_get_string( config, "Session", "screensaver/program", NULL))
    {
        app_command_screensaver(config);
    }

    /* run the power-manager if it's set */
    if (g_key_file_get_string( config, "Session", "power-manager/program", NULL))
    {
        app_command_power_manager(config);
    }

static gboolean load_default_apps( const char* filename )
{
    char buf[1024];
    int len;
    
    FILE* file = fopen( filename, "r" );
    if( !file )
        return FALSE;
    
    while ( fgets( buf, sizeof(buf) - 2, file ) )
    {
        if ( buf[0] == '\0' || buf[0] == '\n' || buf[0] == '#' )
            continue;  /* a comment */
        len = strlen ( buf );
        if( buf[ len - 1 ] == '\n' ) /* remove the '\n' at the end of line */
        {
            --len;
            buf[ len ] = '\0';
        }
        switch(buf[0])
        {
        case '@': /* if the app should be restarted on crash */
            run_app( buf + 1, TRUE );
            break;
        default: /* just run the program */
            run_app( buf, FALSE );
        }
    }
    fclose( file );
    return TRUE;
}

    /* run the file-manager if it's set */
    if (g_key_file_get_string( config, "Session", "file-manager/program", NULL))
    {
        app_command_file_manager(config);
    }

    /* run window manager first from system wide default config */
    if( G_LIKELY( window_manager ) )
        run_app( window_manager, TRUE );

    if( G_UNLIKELY( no_autostart ) )
        return;

    /* load system-wide default apps */
    for( dir = dirs; *dir; ++dir )
    {
        filename = g_build_filename( *dir, prog_name, session_name, autostart_filename, NULL );
        
        /*  g_get_system_config_dirs() may return several directories,
            so ensures that we don't run processes twice. */
        if( load_default_apps( filename ) )
        {
            g_free( filename );
            break;
        }
        g_free( filename );
    }
    
    /* load user-specific default apps */
    filename = g_build_filename( g_get_user_config_dir(), prog_name, session_name, autostart_filename, NULL );
    load_default_apps( filename );
    g_free( filename );

    /* Support autostart spec of freedesktop.org if not disable*/
    xdg_autostart( session_name );

}

static void parse_options(int argc, char** argv)
{
    int  i;
    for ( i = 1; i < argc; ++i )
    {
        if ( argv[i][0] == '-' )
        {
            switch ( argv[i][1] )
            {
            case 'd':     /* -display */
                if ( ++i >= argc ) goto usage;
                display_name = argv[i];
                g_setenv( display_env, display_name, TRUE );
                continue;
            case 's':     /* -session */
                if ( ++i >= argc ) goto usage;
                session_name = argv[i];
                continue;
            case 'n': /* disable xsettings daemon */
				no_settings = TRUE;
                continue;
            case 'e': /* DE name */
                if ( ++i >= argc ) goto usage;
                de_name = argv[i];
                continue;
            case 'r':
				reload_settings = TRUE;
				continue;
            case 'a': /* autostart disable */
                no_autostart = TRUE;
                continue;
			default:
				goto usage;
            }
        }
	}
	return;
usage:
        fprintf ( stderr,
                  "Usage:  lxsession [OPTIONS...]\n"
				  "\t-d NAME\tspecify name of display (optional)\n"
				  "\t-s NAME\tspecify name of the desktop session profile\n"
                  "\t-e NAME\tspecify name of DE, such as LXDE, GNOME, or XFCE.\n"
				  "\t-r\t reload configurations (for Xsettings daemon)\n"
				  "\t-n\t disable Xsettings daemon support\n"
				  "\t-a\t autostart applications disable (window-manager mode only) \n" );
        exit(1);
}

int main(int argc, char** argv)
{
    const char *pid_str;
    char str[ 16 ];
	GKeyFile* kf;
    GObject *clipboard_daemon = NULL;

    gtk_init (&argc, &argv);

    pid_str = g_getenv(pid_env);

    display_name = g_getenv( display_env );
    if( ! display_name )
    {
        display_name = ":0.0";
        g_setenv( display_env, display_name, TRUE );
    }

    parse_options(argc, argv);

    /* initialize X-related stuff and connect to X Display */
    if( G_UNLIKELY(! xevent_init() ) )
        return 1;

    /* send command to existing daemon to reload settings */
    if( G_UNLIKELY( reload_settings ) )
    {
        send_internal_command( LXS_RELOAD );
        return 0;
    }
	else if( G_UNLIKELY( !single_instance_check()) )
	{
		/* only one instance is allowed for each X. */
		fprintf(stderr, "Only one lxsession can be executed at a time");
		return 1;
	}

    /* set pid */
    g_snprintf( str, 16, "%d", getpid() );
    g_setenv(pid_env, str, TRUE );

    main_loop = g_main_loop_new( NULL, TRUE );

	/* setup signal handlers */
    register_signals();

    if ( G_UNLIKELY(!session_name) )
        session_name = "LXDE";
    g_setenv( "DESKTOP_SESSION", session_name, TRUE );

    if ( G_UNLIKELY(!de_name) )
        de_name = session_name;
    g_setenv( "XDG_CURRENT_DESKTOP", de_name, TRUE );

    /* FIXME: load environment variables? */

	/* Load desktop session config file */
	kf = load_session_config(CONFIG_FILE_NAME);
    g_message("Load config file \n");

	if( !kf )
	{
		xevent_finalize();
		return 1;
	}

    if( G_LIKELY(!no_settings) )
        start_settings_daemon(kf);

    if (g_getenv ("XFSETTINGSD_NO_CLIPBOARD") == NULL)
    {
        clipboard_daemon = g_object_new (GSD_TYPE_CLIPBOARD_MANAGER, NULL);
        if (!gsd_clipboard_manager_start (GSD_CLIPBOARD_MANAGER (clipboard_daemon), FALSE))
        {
            g_object_unref (G_OBJECT (clipboard_daemon));
            clipboard_daemon = NULL;

            g_printerr ("Another clipboard manager is already running.");
        }
    }

	g_key_file_free(kf);

    /* start desktop session and load autostart applications */
    g_message("Will start the session \n");
    start_session(kf);

    if (config_changed == TRUE)
    {
        char* rel_path = g_strconcat("lxsession/", session_name, "/desktop.conf", NULL);
        char* dir = g_build_filename(g_get_user_config_dir(), "lxsession/", session_name, NULL);
        char* user_config_file = g_build_filename(g_get_user_config_dir(), rel_path, NULL);
        char* buf;
        gsize len;

        buf = g_key_file_to_data( kf, &len, NULL );

        if (!g_file_test (dir, G_FILE_TEST_IS_DIR))
        {
            g_mkdir_with_parents((dir), 0700);
        }

        g_file_set_contents(user_config_file, buf, len, NULL);
        g_free(buf);
        g_free(dir);
        g_free(user_config_file);
    }

    g_key_file_free(kf);

    g_main_loop_run( main_loop );
    g_main_loop_unref( main_loop );

    if (G_LIKELY (clipboard_daemon != NULL))
    {
        gsd_clipboard_manager_stop (GSD_CLIPBOARD_MANAGER (clipboard_daemon));
        g_object_unref (G_OBJECT (clipboard_daemon));
    }

	xevent_finalize();

    return 0;
}
